CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc

```diff
@@ -601,52 +601,80 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
+	AbilityRanks = GetAbilitiesPerRank(UnitState);
+	bPowerfulAbility = class'X2StrategyGameRulesetDataStructures'.default.PowerfulAbilities.Find(AbilityTree[Branch].AbilityName) != INDEX_NONE;
+	bFinalRankAbility = Rank >= ClassTemplate.GetMaxConfiguredRank() - 1;
+	bXCOMRowAbility = Branch >= AbilityRanks; // This will be "false" for 4th row of abilities for base WOTC Faction Heroes, since they have a randomized deck of perks, not a "true" XCOM row.
+	bAsResistanceHero = UnitState.IsResistanceHero() || AbilityRanks == 0;
 
 	if (GetCustomAbilityCost(UnitState, AbilityTree[Branch].AbilityName, AbilityCost))
 	{
-		return AbilityCost;
+		// Do nothing, AbilityCost will already hold the config value.
 	}
-
-	AbilityRanks = GetAbilitiesPerRank(UnitState);
-	if (!UnitState.IsResistanceHero() && AbilityRanks != 0)
+	else if (!bAsResistanceHero && !UnitState.HasPurchasedPerkAtRank(Rank) && !bXCOMRowAbility)
 	{
-		if (!UnitState.HasPurchasedPerkAtRank(Rank) && Branch < AbilityRanks)
-		{
-			// If this is a base game soldier with a promotion available, ability costs nothing since it would be their
-			// free promotion ability if they "bought" it through the Armory
-			return 0;
-		}
+		// If this is a base game soldier with a promotion available, ability costs nothing.
+		// Do nothing, AbilityCost will already hold 0.
 	}
-
-	bPowerfulAbility = (class'X2StrategyGameRulesetDataStructures'.default.PowerfulAbilities.Find(AbilityTree[Branch].AbilityName) != INDEX_NONE);
-	// Use "poweful ability" cost in following situations:	
-
-	// 1. This is an "AWC" (XCOM row) ability:
-	if (bPowerfulAbility && Branch >= AbilityRanks)
+	else if (bPowerfulAbility && bXCOMRowAbility || bAsResistanceHero && (bPowerfulAbility || bFinalRankAbility))
 	{
+		// Ability Cost is increased for "powerful" abilities in XCOM row of regular soldiers, 
+		// or for "powerful" abilities anywhere in Faction Hero skill tree,
+		// as well as for max rank abilities of Faction Heroes.
 		AbilityCost = class'X2StrategyGameRulesetDataStructures'.default.PowerfulAbilityPointCost;
 	}
-
-	// 2. All Colonel level abilities for emulated Faction Heroes and any powerful XCOM abilities have increased cost for Faction Heroes
-	if (AbilityRanks == 0 && (bPowerfulAbility || (Rank >= 6 && Branch < 3)) && !HasBrigadierRank())
+	else
 	{
-		return class'X2StrategyGameRulesetDataStructures'.default.PowerfulAbilityPointCost;
-	}
+		AbilityCost = GetDefaultAbilityPointCostForRank(Rank);
+	}	
 
-	// 3. All Colonel level abilities for Faction Heroes and any powerful XCOM abilities have increased cost for Faction Heroes
-	if (UnitState.IsResistanceHero() && (bPowerfulAbility || (Rank >= 6 && Branch < 3)) && !HasBrigadierRank())
+	// Start Issue #10
+	/// Mods can listen to 'OverrideAbilityPointCost' event to use their own logic 
+	/// to determine abiility point cost for this particular unit and this particular ability.
+	/// It can be used to reduce ability point cost to 0, if necessary.
```

Comment:
Probably worth clarifying that `Rank` and `Row` are indexed from zero just for complete clarity. It's a fair assumption that they're indexed from 0, but it's worth removing any element of doubt.

Reply:
I would also clarify (if this is the case) that `Rank` does not refer to the normal `XCGS_Unit` rank, which has 0 for Rookie. I think in this case 0 is for squaddie, right? I'm being a bit lazy in not working it out, but I'm running out of steam.

Reply:
I'll investigate and leave the appropriate note.